packages/wrangler/src/d1/execute.tsx

```diff
@@ -348,41 +346,168 @@ async function executeRemotely({
-			);
+	if (input.file) {
+		// TODO: chunk into multipart for upload speed
+		const etag = generateETag(input.file);
+
+		logger.log(
+			chalk.gray(
+				`Note: if the execution fails to complete, your DB will return to its original state and you can safely retry.`
+			)
+		);
+
+		const initResponse = await d1ApiPost<
+			| ImportInitResponse
+			| ImportPollingResponse
+			| { success: false; error: string }
+		>(accountId, db, "import", { action: "init", etag });
+		console.log(initResponse);
 
-		const result = await fetchResult<QueryResult[]>(
-			`/accounts/${accountId}/d1/database/${dbUuid}/query`,
-			{
-				method: "POST",
-				headers: {
-					"Content-Type": "application/json",
-					...(db.internal_env ? { "x-d1-internal-env": db.internal_env } : {}),
-				},
-				body: JSON.stringify({ sql }),
-			}
+		// An init response usually returns a {filename, uploadUrl} pair, except if we've detected that file
+		// already exists and is valid, to save people reuploading. In which case `initResponse` has already
+		// kicked the import process off.
+		const firstPollResponse =
+			"uploadUrl" in initResponse
+				? await uploadAndBeginIngestion(
+						accountId,
+						db,
+						input.file,
+						etag,
+						initResponse
+				  )
+				: initResponse;
+
+		const finalResponse = await pollUntilComplete(
+			firstPollResponse,
+			accountId,
+			db
 		);
+
+		if (finalResponse.status !== "complete")
+			throw new APIError({ text: `D1 reset before execute completed!` });
+
+		logger.log(`Done!`);
+
+		return null;
+	} else {
+		const result = await d1ApiPost<QueryResult[]>(accountId, db, "query", {
+			sql: input.command,
+		});
 		logResult(result);
-		results.push(...result);
+		return result;
+	}
+}
+
+async function uploadAndBeginIngestion(
+	accountId: string,
+	db: Database,
+	file: string,
+	etag: string,
+	initResponse: ImportInitResponse
+) {
+	const { uploadUrl, filename } = initResponse;
+	logger.log(`🌀 Uploading ${filename}...`);
+
+	const { size } = await fs.stat(file);
+
+	const uploadResponse = await fetch(uploadUrl, {
+		method: "PUT",
+		headers: {
+			"Content-length": `${size}`,
+		},
+		body: createReadStream(file),
+		duplex: "half", // required for NodeJS streams over .fetch ?
+	});
+
+	if (uploadResponse.status !== 200) {
+		console.log({ uploadResponse });
+		console.log(await uploadResponse.text());
+		throw new UserError(`File could not be uploaded. Please retry.`);
+	}
+
+	const etagResponse = uploadResponse.headers.get("etag");
+	if (!etagResponse) {
+		throw new UserError(`File did not upload successfully. Please retry.`);
+	}
+	console.log({ etag, etagResponse });
+	if (etag !== etagResponse.replace(/^"|"$/g, "")) {
+		throw new UserError(
+			`File contents did not upload successfully. Please retry.`
+		);
 	}
-	return results;
+	logger.log(`🌀 Uploading complete.`);
+
+	return await d1ApiPost<
+		ImportPollingResponse | { success: false; error: string }
```

Comment:
we repeat `{ success: false; error: string }` a few times, could be worth having its own type